workspaces/kubernetes-native-dxp/modules/kubernetes-agent/src/main/java/com/liferay/k8s/agent/internal/K8sAgentImpl.java

```diff
@@ -0,0 +1,438 @@
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import org.osgi.service.cm.Configuration;
+import org.osgi.service.cm.ConfigurationAdmin;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.ConfigurationPolicy;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Reference;
+
+import com.liferay.k8s.agent.K8sAgent;
+import com.liferay.k8s.agent.configuration.v1.K8sAgentConfiguration;
+import com.liferay.k8s.agent.properties.ConfigurationProperties;
+import com.liferay.k8s.agent.properties.ConfigurationPropertiesFactory;
+import com.liferay.petra.string.CharPool;
+import com.liferay.petra.string.StringBundler;
+import com.liferay.petra.string.StringPool;
+import com.liferay.portal.configuration.metatype.bnd.util.ConfigurableUtil;
+import com.liferay.portal.kernel.log.Log;
+import com.liferay.portal.kernel.log.LogFactoryUtil;
+import com.liferay.portal.kernel.util.GetterUtil;
+import com.liferay.portal.kernel.util.HashMapDictionary;
+import com.liferay.portal.util.PropsValues;
+
+import io.fabric8.kubernetes.api.model.ConfigMap;
+import io.fabric8.kubernetes.api.model.ConfigMapBuilder;
+import io.fabric8.kubernetes.api.model.ObjectMeta;
+import io.fabric8.kubernetes.client.Config;
+import io.fabric8.kubernetes.client.DefaultKubernetesClient;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.Watch;
+import io.fabric8.kubernetes.client.Watcher;
+import io.fabric8.kubernetes.client.Watcher.Action;
+import io.fabric8.kubernetes.client.WatcherException;
+
+/**
+ * @author Raymond Augé
+ */
+@Component(
+	configurationPid = "com.liferay.k8s.agent.configuration.v1.K8sAgentConfiguration",
+	configurationPolicy = ConfigurationPolicy.REQUIRE,
+	immediate = true,
+	service = K8sAgent.class
+)
+public class K8sAgentImpl implements K8sAgent {
+
+	@Activate
+	public K8sAgentImpl(
+			@Reference ConfigurationAdmin configurationAdmin,
+			Map<String, Object> properties)
+		throws Exception {
+
+		_configurationAdmin = configurationAdmin;
+
+		_k8sAgentConfiguration =
+			ConfigurableUtil.createConfigurable(
+				K8sAgentConfiguration.class, properties);
+
+		_executor = Executors.newSingleThreadExecutor();
+
+		_log.info(
+			StringBundler.concat(
+				"Initializing ", _AGENT_NAME, ": ",
+				_k8sAgentConfiguration.namespace(), " ",
+				_k8sAgentConfiguration.labelSelector()
+			)
+		);
+
+		Config config = Config.autoConfigure(null);
+
+		_kubernetesClient = new DefaultKubernetesClient(config);
+
+		_watch = _kubernetesClient.configMaps(
+		).inNamespace(
+			_k8sAgentConfiguration.namespace()
+		).withLabel(
+			_k8sAgentConfiguration.labelSelector()
+		).watch(
+			new Watcher<ConfigMap>() {
```

Comment:
@rotty3000, I wonder if we should refactor this later to use Informer pattern instead of the plain watch.
I found a few references like this: https://rohaan.medium.com/introduction-to-fabric8-kubernetes-java-client-informer-api-b945082d69af

Reply:
possibly, I'll look into it

Reply:
I've made the change to informer, will commit by end of day.

Reply:
@rotty3000 please tag me once you've committed this change. Thanks

Reply:
@igorarouca it's probably better for you to just follow our branch here https://github.com/rotty3000/liferay-portal/tree/workspaces